overlord/snapstate/handlers_discard_test.go

```diff
@@ -82,6 +88,52 @@ func (s *discardSnapSuite) TestDoDiscardSnapSuccess(c *C) {
 	c.Check(t.Status(), Equals, state.DoneStatus)
 }
 
+func (s *discardSnapSuite) TestDoDiscardSnapInQuotaGroup(c *C) {
+	s.state.Lock()
```

Comment:
(idle-though) I dislike the "Lock(); random-lines; Unlock()" we often (have to) do in the tests. I wonder if there is a better pattern, crazy ideas would be: 
```
var t *state.Task
st.Lock()
{
     // do all the stuff
}
st.Unlock()
```
Not for this PR and maybe never, mostly idle wondering and maybe not even a big deal given that these are tests and it's usually short enough to see where we lock/unlock. Anyway, sorry for ramping a bit here.

Reply:
personally when we have to do that a lot, I like this pattern better:

```go
func foo(st *state.State) {
    st.Lock()
    defer st.Unlock()
    // do the things with the lock

    st.Unlock()
    defer st.Lock()
    // do the foo things that need to be unlocked

    st.Lock()
    defer st.Unlock()
    // do the other foo things that need to be locked again
}

Reply:
@anonymouse64, are you sure you want to defer a `st.Lock()`? Wouldn't that cause the lock to be held forever, if the function returns while executing the code in the ` // do the foo things that need to be unlocked` part?

Reply:
@mardy anything that's is deferred is guaranteed to be run when the function returns (that's the point of `defer` in go), so this is correct and we do that a lot.

Reply:
Yes, I understand that, but while I understand why you'd want to defer an `Unlock()`, it's harder to me to imagine a case where deferring a `Lock()` would be useful (and not harmful).

Reply:
@mardy the order in which defer'd functions is ordered is like a stack, so the most recently defer'd one is executed first, so if we exited the function after the second defer, then we would first execute st.Lock() (which is okay since we exited the function with the lock not held), then we execute the first defer, to unlock it again, which is okay since we just acquired the lock. I admit it can be a bit confusing to read at first but the go language specification says that defer'd functions are executed in this order.

Reply:
mmm... OK, this part is clear now, thanks. Still, I fail to understand *why* we would want to do that. In other words, if the code was
```golang
func foo(st *state.State) {
    st.Lock()
    defer st.Unlock()
    // do the things with the lock

    st.Unlock()
    // do the foo things that need to be unlocked

    st.Lock()
    defer st.Unlock() // <- at this point this could be omitted, I guess
    // do the other foo things that need to be locked again
}
```
that is, exactly the same as you wrote, but without the deferred `Lock()`, what would change?

Reply:
@mardy in the example snippet you provided, your program would panic after returning from the function since you cannot unlock an already unlocked mutex. It would not panic if you left out the `st.Unlock()` at the end, but it suffers from the problem that if you happened to return at the line `// do the foo things that need to be unlocked` then you would also panic since your defered `st.Unlock()` would be unlocked an unlocked mutex which panics.